Enable passing retryable `SqlLiteral`s to `#where`

skipkayhil opened the pull request:
### Motivation / Background

Previously, retryable SqlLiterals passed to `#where` would lose their retryability because both `#build_where_clause` and `WhereClause` would wrap them in non-retryable `SqlLiteral`s.

### Detail

To fix this problem, this commit updates `#build_where_clause` to check for `SqlLiterals` and updates `WhereClause` to assume that any predicate passed in will already be wrapped.

### Additional information

There were only a few places that passed Strings to `WhereClause` (`"1=0"` in `PredicateBuilder` and sanitized sql strings in `#build_where_clause`) that needed to be updated to use `Arel.sql`. The `WhereClause` tests also had Strings to wrap but they are really unit tests and aren't representative of what can be done with the public API

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

skipkayhil committed:
Enable passing retryable `SqlLiteral`s to `#where`

Previously, retryable SqlLiterals passed to `#where` would lose their
retryability because both `#build_where_clause` and `WhereClause` would
wrap them in non-retryable `SqlLiteral`s.

To fix this problem, this commit updates `#build_where_clause` to check
for `SqlLiterals` and updates `WhereClause` to assume that any predicate
passed in will already be wrapped.

There were only a few places that passed Strings to `WhereClause`
(`"1=0"` in `PredicateBuilder` and sanitized sql strings in
`#build_where_clause`) that needed to be updated to use `Arel.sql`. The
`WhereClause` tests also had Strings to wrap but they are really unit
tests and aren't representative of what can be done with the public API

skipkayhil commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -1620,13 +1620,17 @@ def build_where_clause(opts, rest = []) # :nodoc:
             parts = [build_bound_sql_literal(opts, rest)]
           else
-            parts = [model.sanitize_sql(rest.empty? ? opts : [opts, *rest])]
+            parts = [Arel.sql(model.sanitize_sql([opts, *rest]))]
```
The `rest.empty?` check appears to have been dead code since 8e6a5deca626c944933211d7977f4e7e76869e74

byroot commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -1620,13 +1620,17 @@ def build_where_clause(opts, rest = []) # :nodoc:
         when String
           if rest.empty?
-            parts = [Arel.sql(opts)]
+            if Arel.arel_node?(opts)
+              parts = [opts]
+            else
+              parts = [Arel.sql(opts)]
+            end
```
Shouldn't we just do that in `Arel.sql`? Right now it assumes it always receives a `String`, hence always wrap it, but we could make it return arel nodes unchanged.

skipkayhil replied:
I tried searching for usages of `Arel.sql` that would pass a `SqlLiteral`, but it seems like there aren't many (any?) others than this one (I used `❯ rg '[^rf]\.sql\([^"%]'`).

I think it could make sense to do if the pattern was more common, but it seems like 90+% of the time a string is passed anyways.

skipkayhil replied:
For a more definitive answer, I ran `ARCONN=sqlite3 bin/test` with this patch:

```
diff --git a/activerecord/lib/arel.rb b/activerecord/lib/arel.rb
index 738e80df35..cf82e60656 100644
--- a/activerecord/lib/arel.rb
+++ b/activerecord/lib/arel.rb
@@ -50,6 +50,7 @@ module Arel
   # Use this option only if the SQL is idempotent, as it could be executed
   # more than once.
   def self.sql(sql_string, *positional_binds, retryable: false, **named_binds)
+    raise if Nodes::SqlLiteral === sql_string
     if positional_binds.empty? && named_binds.empty?
       Arel::Nodes::SqlLiteral.new(sql_string, retryable: retryable)
     else
```

It fails on main but passes on this branch, so the two places touched by this PR are the only places we pass a `SqlLiteral` to `Arel.sql`.

byroot replied:
I guess my point was more for the future, and for general API design. e.g. `Array([]) # => []`.

matthewd replied:
I feel a bit non-specifically wary of defining `Arel.sql` to pass-through all Nodes (while conceding that's functionally what we do in several call-sites already)... but regardless of that, I do agree it seems reasonable/sensible for it to pass-through its own results rather than re-wrapping them.

skipkayhil replied:
Updated to pass through `SqlLiterals` 👍 I agree passing through other Nodes probably doesn't make sense (and I think `Arel.sql` doesn't even accept other nodes at the moment since they can't be coerced to String).

skipkayhil committed:
Remove SqlLiteral re-wrapping from Arel.sql

Previously, passing a SqlLiteral to Arel.sql would wrap the SqlLiteral
in a new SqlLiteral (which is generally unnecessary since it allocates a
new SqlLiteral/String).

This commit updates Arel.sql to no longer perform the wrapping of the
given "sql_string" is already a SqlLiteral.

skipkayhil commented:
> @skipkayhil thanks for your awesome work on this! Will `BoundSqlLiteral`s also be supported soon?

There's a linked PR just above your comment, but I can't speak to if/when it'll be merged
